agent/consul/acl_endpoint_test.go

```diff
@@ -806,184 +875,1952 @@ func TestACLEndpoint_TokenSet(t *testing.T) {
+				ACLToken: structs.ACLToken{
+					Description:   "foobar",
+					Policies:      nil,
+					Local:         false,
+					ExpirationTTL: test.offset,
+				},
+				WriteRequest: structs.WriteRequest{Token: "root"},
+			}
+
+			resp := structs.ACLToken{}
+
+			err := acl.TokenSet(&req, &resp)
+			if test.errString != "" {
+				requireErrorContains(t, err, test.errStringTTL)
+			} else {
+				require.NotNil(t, err)
+			}
+		})
+	}
+
+	t.Run("Create it with expiration time AND expiration TTL set (error)", func(t *testing.T) {
+		req := structs.ACLTokenSetRequest{
+			Datacenter: "dc1",
+			ACLToken: structs.ACLToken{
+				Description:    "foobar",
+				Policies:       nil,
+				Local:          false,
+				ExpirationTime: time.Now().Add(4 * time.Second),
+				ExpirationTTL:  4 * time.Second,
+			},
+			WriteRequest: structs.WriteRequest{Token: "root"},
+		}
+
+		resp := structs.ACLToken{}
+
+		err := acl.TokenSet(&req, &resp)
+		requireErrorContains(t, err, "Expiration TTL and Expiration Time cannot both be set")
+	})
+
+	t.Run("Create it with expiration time using TTLs", func(t *testing.T) {
+		req := structs.ACLTokenSetRequest{
+			Datacenter: "dc1",
+			ACLToken: structs.ACLToken{
+				Description:   "foobar",
+				Policies:      nil,
+				Local:         false,
+				ExpirationTTL: 4 * time.Second,
+			},
+			WriteRequest: structs.WriteRequest{Token: "root"},
+		}
+
+		resp := structs.ACLToken{}
+
+		err := acl.TokenSet(&req, &resp)
+		require.NoError(t, err)
+
+		// Get the token directly to validate that it exists
+		tokenResp, err := retrieveTestToken(codec, "root", "dc1", resp.AccessorID)
+		require.NoError(t, err)
+		token := tokenResp.Token
+
+		expectExpTime := resp.CreateTime.Add(4 * time.Second)
+
+		require.NotNil(t, token.AccessorID)
+		require.Equal(t, token.Description, "foobar")
+		require.Equal(t, token.AccessorID, resp.AccessorID)
+		requireTimeEquals(t, expectExpTime, resp.ExpirationTime)
+
+		tokenID = token.AccessorID
+	})
+
+	var expTime time.Time
+	t.Run("Create it with expiration time", func(t *testing.T) {
+		expTime = time.Now().Add(4 * time.Second)
+		req := structs.ACLTokenSetRequest{
+			Datacenter: "dc1",
+			ACLToken: structs.ACLToken{
+				Description:    "foobar",
+				Policies:       nil,
+				Local:          false,
+				ExpirationTime: expTime,
+			},
+			WriteRequest: structs.WriteRequest{Token: "root"},
+		}
+
+		resp := structs.ACLToken{}
+
+		err := acl.TokenSet(&req, &resp)
+		require.NoError(t, err)
+
+		// Get the token directly to validate that it exists
+		tokenResp, err := retrieveTestToken(codec, "root", "dc1", resp.AccessorID)
+		require.NoError(t, err)
+		token := tokenResp.Token
+
+		require.NotNil(t, token.AccessorID)
+		require.Equal(t, token.Description, "foobar")
+		require.Equal(t, token.AccessorID, resp.AccessorID)
+		requireTimeEquals(t, expTime, resp.ExpirationTime)
+
+		tokenID = token.AccessorID
+	})
+
+	// do not insert another test at this point: these tests need to be serial
```

Comment:
Why put them in separate `t.Run` calls? You can just wrap them in braces if you want to limit scope and re-use variables etc we do that in a few places.

If someone added `t.Parallel` to this test at the top for example this would be violated too.